aws/rust-runtime/aws-config/src/environment/retry_config.rs

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+use std::str::FromStr;
+
+use aws_types::os_shim_internal::Env;
+use smithy_types::retry::{RetryConfig, RetryMode};
+
+const ENV_VAR_MAX_ATTEMPTS: &str = "AWS_MAX_ATTEMPTS";
+const ENV_VAR_RETRY_MODE: &str = "AWS_RETRY_MODE";
+
+/// Load a retry_config from environment variables
+///
+/// This provider will check the values of `AWS_RETRY_MODE` and `AWS_MAX_ATTEMPTS`
+/// in order to build a retry config. If at least one is set to a valid value,
+/// construction will succeed
+#[derive(Debug, Default)]
+pub struct EnvironmentVariableRetryConfigProvider {
+    env: Env,
+}
+
+impl EnvironmentVariableRetryConfigProvider {
+    /// Create a new `EnvironmentVariableRetryConfigProvider`
+    pub fn new() -> Self {
+        EnvironmentVariableRetryConfigProvider { env: Env::real() }
+    }
+
+    #[doc(hidden)]
+    /// Create an retry_config provider from a given `Env`
+    ///
+    /// This method is used for tests that need to override environment variables.
+    pub fn new_with_env(env: Env) -> Self {
+        EnvironmentVariableRetryConfigProvider { env }
+    }
+
+    /// Attempt to create a new `RetryConfig` from environment variables
+    pub fn retry_config(&self) -> Option<RetryConfig> {
```

Comment:
Personally, I think I'd have this function return a `Result<Option<RetryConfig>, InvalidRetryConfig>` and save the panic for the top level. I don't love that the panic will be hidden deep in the code.